### ISSUE TITLE
Fails to compile with ASIO_NO_TYPEID,  ASIO_ENABLE_HANDLER_TRACKING

### ISSUE THREAD

**Issue by byllyfish** [chriskohlhoff/asio#50]
I'm using the latest ASIO 1.11. When I define ASIO_NO_TYPEID and ASIO_ENABLE_HANDLER_TRACKING there are compilation errors.

The following changes appear to mitigate the errors.

```diff
diff --git a/asio/include/asio/detail/impl/strand_executor_service.hpp b/asio/include/asio/detail/impl/strand_exe
index 20cc24c..c518433 100644
--- a/asio/include/asio/detail/impl/strand_executor_service.hpp
+++ b/asio/include/asio/detail/impl/strand_executor_service.hpp
@@ -121,7 +121,7 @@ void strand_executor_service::dispatch(const implementation_type& impl,
   p.v = p.a.allocate(1);
   p.p = new (p.v) op(tmp, allocator);
 
-  ASIO_HANDLER_CREATION((p.p, "strand_executor", this, "dispatch"));
+  ASIO_HANDLER_CREATION((p.p, "strand_executor", &impl, "dispatch"));
 
   // Add the function to the strand and schedule the strand if required.
   bool first = enqueue(impl, p.p);
@@ -149,7 +149,7 @@ void strand_executor_service::post(const implementation_type& impl,
   p.v = p.a.allocate(1);
   p.p = new (p.v) op(tmp, allocator);
 
-  ASIO_HANDLER_CREATION((p.p, "strand_executor", this, "post"));
+  ASIO_HANDLER_CREATION((p.p, "strand_executor", &impl, "post"));
 
   // Add the function to the strand and schedule the strand if required.
   bool first = enqueue(impl, p.p);
@@ -177,7 +177,7 @@ void strand_executor_service::defer(const implementation_type& impl,
   p.v = p.a.allocate(1);
   p.p = new (p.v) op(tmp, allocator);
 
-  ASIO_HANDLER_CREATION((p.p, "strand_executor", this, "defer"));
+  ASIO_HANDLER_CREATION((p.p, "strand_executor", &impl, "defer"));
 
   // Add the function to the strand and schedule the strand if required.
   bool first = enqueue(impl, p.p);
diff --git a/asio/include/asio/executor.hpp b/asio/include/asio/executor.hpp
index a3b5730..2dec890 100644
--- a/asio/include/asio/executor.hpp
+++ b/asio/include/asio/executor.hpp
@@ -205,10 +205,12 @@ public:
    * @returns If @c *this has a target type of type @c T, <tt>typeid(T)</tt>;
    * otherwise, <tt>typeid(void)</tt>.
    */
+#if !defined(ASIO_NO_TYPEID)
   const std::type_info& target_type() const ASIO_NOEXCEPT
   {
     return impl_ ? impl_->target_type() : typeid(void);
   }
+#endif // !defined(ASIO_NO_TYPEID)
 
   /// Obtain a pointer to the target executor object.
   /**
diff --git a/asio/include/asio/impl/executor.hpp b/asio/include/asio/impl/executor.hpp
index d8bc403..40f7799 100644
--- a/asio/include/asio/impl/executor.hpp
+++ b/asio/include/asio/impl/executor.hpp
@@ -197,10 +197,12 @@ public:
     executor_.defer(ASIO_MOVE_CAST(function)(f), allocator_);
   }
 
+#if !defined(ASIO_NO_TYPEID)
   const std::type_info& target_type() const ASIO_NOEXCEPT
   {
     return typeid(Executor);
   }
+#endif // !defined(ASIO_NO_TYPEID)
 
   void* target() ASIO_NOEXCEPT
   {
@@ -306,10 +308,12 @@ public:
     executor_.defer(ASIO_MOVE_CAST(function)(f), allocator_);
   }
 
+#if !defined(ASIO_NO_TYPEID)
   const std::type_info& target_type() const ASIO_NOEXCEPT
   {
     return typeid(system_executor);
   }
+#endif // !defined(ASIO_NO_TYPEID)
 
   void* target() ASIO_NOEXCEPT
   {
@@ -365,6 +369,8 @@ void executor::defer(ASIO_MOVE_ARG(Function) f, const Allocator& a)
   get_impl()->defer(function(ASIO_MOVE_CAST(Function)(f), a));
 }
 
+#if !defined(ASIO_NO_TYPEID)
+
 template <typename Executor>
 Executor* executor::target() ASIO_NOEXCEPT
 {
@@ -379,6 +385,8 @@ const Executor* executor::target() const ASIO_NOEXCEPT
     ? static_cast<Executor*>(impl_->target()) : 0;
 }
 
+#endif // !defined(ASIO_NO_TYPEID)
+
 #endif // !defined(ASIO_NO_DEPRECATED)
 
 } // namespace asio
```
On an unrelated note, the `#endif // !defined(ASIO_NO_DEPRECATED)` near the end of the diff corresponds to a `#if !defined(GENERATING_DOCUMENTATION)` at the top of the file.

**Comment by byllyfish** [comment]
There are also issues with ASIO_NO_EXCEPTION.

```diff
diff --git a/asio/include/asio/executor.hpp b/asio/include/asio/executor.hpp
index a3b5730..ec1e0a6 100644
--- a/asio/include/asio/executor.hpp
+++ b/asio/include/asio/executor.hpp
@@ -278,7 +280,8 @@ private:
   // Helper function to check and return the implementation pointer.
   impl_base* get_impl()
   {
-    return impl_ ? impl_ : throw bad_executor();
+    if (!impl_) asio::detail::throw_exception(bad_executor());
+    return impl_;
   }
 
   // Helper function to clone another implementation.

diff --git a/asio/include/asio/ip/address.hpp b/asio/include/asio/ip/address.hpp
index 7cddee2..b8dd909 100644
--- a/asio/include/asio/ip/address.hpp
+++ b/asio/include/asio/ip/address.hpp
@@ -242,7 +242,7 @@ inline T address_cast(const address& addr,
     typename enable_if<is_same<T, address_v4>::value>::type* = 0)
 {
   if (!addr.is_v4())
-    throw bad_address_cast();
+    asio::detail::throw_exception(bad_address_cast());
   return get_v4_helper(addr);
 }
 
@@ -255,7 +255,7 @@ inline T address_cast(const address& addr,
     typename enable_if<is_same<T, address_v6>::value>::type* = 0)
 {
   if (!addr.is_v6())
-    throw bad_address_cast();
+    asio::detail::throw_exception(bad_address_cast());
   return get_v6_helper(addr);
 }
```

**Comment by armandf** [comment]
Thanks a lot for these fixes, mr. byllyfish, you saved me some time. I hope the changes get merged soon. I am relying on ASIO_NO_TYPEID and ASIO_NO_EXCEPTION.

**Comment by chriskohlhoff** [comment]
Fixed in various commits, including df9d70bc553e81c17f2be5f8010d4d4e920e8731 and 23b084bb617f91ab0bbeddc796cf72d98663f236.
